lib/common/rules.c

```diff
@@ -1319,12 +1319,7 @@ pcmk__evaluate_condition(xmlNode *expr, const pcmk_rule_input_t *rule_input,
 
         case pcmk__subexpr_attribute:
         case pcmk__subexpr_location:
-            /* these expressions can never succeed if there is
-             * no node to compare with
-             */
-            if (rule_input->node_attrs != NULL) {
-                rc = pcmk__evaluate_attr_expression(expr, rule_input);
-            }
+            rc = pcmk__evaluate_attr_expression(expr, rule_input);
```

Comment:
I consider it unsupported to create a node object without using an API call (which guarantees that the attribute table is created). I'm not sure that's plainly documented anywhere, but once we make the objects opaque, it'll be guaranteed.

Node attribute expressions are for a specified attribute, so the attribute table can have other entries and still pass an undefined test for that attribute.

Reply:
I mean something could create an arbitrary `pcmk_rule_input_t` object with any value (or NULL) for the `node_attrs` member, and call `pcmk__evaluate_condition()` against an attribute expression. At that point, node object creation doesn't come into play. We usually don't make sanity assumptions about the call chain except in one-call statics.

It's not a big deal if we do here; however, I wanted to point it out.

Reply:
A NULL `node_attrs` is fine -- the only issue here is that it's being used to determine whether to evaluate a node attribute expression, which it shouldn't. That's more of a bug than a behavior. It's irrelevant for internal callers since they will always pass a non-NULL `node_attrs` when a node attribute expression is allowed.